src/common/metrics.rs

```diff
@@ -806,138 +806,229 @@ fn label_pair(name: &str, value: &str) -> LabelPair {
+                vec![counter(major_page_faults as f64, &[])],
+                prefix,
+            ));
+        }
     }
-}
 
-#[cfg(target_os = "linux")]
-impl MetricsProvider for ProcFsMetrics {
-    fn add_metrics(&self, metrics: &mut Vec<MetricFamily>, prefix: Option<&str>) {
-        metrics.push(metric_family(
-            "process_threads",
-            "count of active threads",
-            MetricType::GAUGE,
-            vec![gauge(self.thread_count as f64, &[])],
-            prefix,
-        ));
+    /// Returns the minor and major page faults of all children including grandchildren, etc.
+    pub fn faults_for_all_children(parent: Pid) -> Result<(u64, u64), ProcError> {
+        let children = child_processes_helper(parent)?;
+        let mut min = 0;
+        let mut maj = 0;
+        for child in children {
+            let stat = match Process::new(child).and_then(|i| i.stat()) {
+                Ok(proc) => proc,
+                // Ignore children that don't exist anymore.
+                Err(ProcError::NotFound(_)) => continue,
+                Err(err) => return Err(err),
+            };
 
-        metrics.push(metric_family(
-            "process_open_mmaps",
-            "count of open mmaps",
-            MetricType::GAUGE,
-            vec![gauge(self.mmap_count as f64, &[])],
-            prefix,
-        ));
+            min += stat.minflt;
+            maj += stat.majflt;
+        }
+        Ok((min, maj))
+    }
 
-        metrics.push(metric_family(
-            "system_max_mmaps",
-            "system wide limit of open mmaps",
-            MetricType::GAUGE,
-            vec![gauge(self.system_mmap_limit as f64, &[])],
-            prefix,
-        ));
+    /// Returns a list of all children of a given process. This works recursively, including grandchildren, etc.
+    fn child_processes_helper(pid: Pid) -> Result<Vec<Pid>, ProcError> {
+        let mut seen = HashSet::default();
+        seen.insert(pid); // Exclude parent Pid from results
 
-        metrics.push(metric_family(
-            "process_open_fds",
-            "count of currently open file descriptors",
-            MetricType::GAUGE,
-            vec![gauge(self.open_fds as f64, &[])],
-            prefix,
-        ));
+        let mut out = HashSet::default();
 
-        let fds_limit = match (self.max_fds_soft, self.max_fds_hard) {
-            (0, hard) => hard,                         // soft unlimited, use hard
-            (soft, 0) => soft,                         // hard unlimited, use soft
-            (soft, hard) => std::cmp::min(soft, hard), // both limited, use minimum
-        };
-        metrics.push(metric_family(
-            "process_max_fds",
-            "limit for open file descriptors",
-            MetricType::GAUGE,
-            vec![gauge(fds_limit as f64, &[])],
-            prefix,
-        ));
+        child_processes_recursive(pid, &mut seen, &mut out)?;
+        Ok(out.into_iter().collect())
```

Comment:
We can probably use a single hashset here. But it's not super important.

Reply:
Done in https://github.com/qdrant/qdrant/pull/7482/commits/faadcc315d716226dadd7cab9f933cdca20cdbf2

We can use the returned boolean from [`HashSet::insert(...)`](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.insert) to determine if we've seen a PID or not.